beacon-chain/p2p/subnets.go

```diff
@@ -100,94 +102,127 @@ func (s *Service) FindPeersWithSubnets(
 
 	defectiveSubnets := s.defectiveSubnets(topicFormat, digest, minimumPeersPerSubnet, subnets)
 	for len(defectiveSubnets) > 0 {
-		if err := func() error {
-			// Create an discovery iterator to find new peers.
-			iterator := s.dv5Listener.RandomNodes()
-
-			// The peer crawling stops when enough peers are found in all defective subnets,
-			// or when `batchPeriod` is reached.
-			// It is useful to dial peers even if not all peers in defective subnets
-			// are found.
+		// Stop the search/dialing loop if the context is canceled.
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		peersToDial, err := func() ([]*enode.Node, error) {
 			ctx, cancel := context.WithTimeout(ctx, batchPeriod)
 			defer cancel()
 
-			// `iterator.Next` can block indefinitely. `iterator.Close` unblocks it.
-			// So it is important to close the iterator when the context is done to ensure
-			// that the search does not hang indefinitely.
-			go func() {
-				<-ctx.Done()
-				iterator.Close()
-			}()
-
-			// Retrieve the filter function that will be used to filter nodes based on the defective subnets.
-			filter, err := s.nodeFilter(topicFormat, defectiveSubnets)
-			if err != nil {
-				return errors.Wrap(err, "node filter")
+			peersToDial, err := s.findPeersWithSubnets(ctx, topicFormat, digest, minimumPeersPerSubnet, defectiveSubnets)
+			if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+				return nil, errors.Wrap(err, "find peers with subnets")
 			}
 
-			// Crawl the network for peers subscribed to the defective subnets.
-			nodeByNodeID := make(map[enode.ID]*enode.Node)
-			for ctx.Err() == nil && len(defectiveSubnets) > 0 && iterator.Next() {
-				// Get all needed subnets that the node is subscribed to.
-				// Skip nodes that are not subscribed to any of the defective subnets.
-				node := iterator.Node()
-				nodeSubnets, err := filter(node)
-				if err != nil {
-					return errors.Wrap(err, "filter node")
-				}
-				if len(nodeSubnets) == 0 {
-					continue
-				}
+			return peersToDial, nil
+		}()
 
-				// Remove duplicates, keeping the node with higher seq.
-				existing, ok := nodeByNodeID[node.ID()]
-				if ok && existing.Seq() > node.Seq() {
-					continue
-				}
-				nodeByNodeID[node.ID()] = node
+		if err != nil {
+			return errors.Wrap(err, "find peers with subnets batch")
```

Comment:
similar comment as previous one

Reply:
Fixed in 35ee1f2df57ec9a5837bff6b0aa2e3c39409c023.
In this case, no need to wrap.